File: math/prime/prime.py
import math 
from typing import List

def IsPrimeBruteForce (n : int) -> bool:

    for i in range(2,n):
        if n % i == 0:
            return False
    return True
    
def IsPrime(n : int) -> bool:

  if n == 2:
      return True
  if n <= 1 or n % 2 == 0:
      return False
  
  m = int(math.sqrt(n))

  for i in range(3,m,2):
        if n % i == 0:
            return False

  return True
  
def Sieve(n : int) -> List[bool]:

  prime = [True] * (n+1)
  prime[0] = False
  prime[1] = False

  m = int(math.sqrt(n))

  for i in range(2,m+1):
      if prime[i]:
        for j in range(i**2,n+1,i):
            prime[j] = False
            
  return prime

File: math/prime.cpp
#include <cmath>
#include <vector>

using namespace std;

bool IsPrimeBruteForce (int n) {

    for (int i = 2; i < n; i++) {
        if (n % i == 0) {
            return false;
        }
    }

    return true;
}

bool IsPrime(int n) {

  if (n == 2) return true;
  if (n <= 1 || n % 2 == 0) return false;
  
  int m = sqrt(n);

  for (int i = 3; i <= m; i += 2) {
        if (n % i == 0) {
            return false;
        }
  }

  return true;
}

vector<bool> Sieve(int n) {

  vector<bool> prime(n + 1, true);
  prime[0] = false;
  prime[1] = false;

  int m = sqrt(n);

  for (int i = 2; i <= m; i++) {
      if (prime[i]) {
        for (int j = i * i; j <= n; j += i) {
            prime[j] = false;
        }           
    }
  }
    
  return prime;
}

